mini-projects/00-jq-rs/src/filters.rs

```diff
@@ -0,0 +1,278 @@
+    }
+}
+
+fn array_slice(input: &Value, index_str: &str) -> Result<Value, MyErrors> {
+    let indices: Vec<&str> = index_str.split(':').collect();
+    let start: usize = indices[0].parse().map_err(MyErrors::ParseError)?;
+    let end: usize = indices[1].parse().map_err(MyErrors::ParseError)?;
+
+    if start < end && end <= input.as_array().unwrap().len() {
+        let slice = &input.as_array().unwrap()[start..end];
+        Ok(Value::Array(slice.to_vec()))
+    } else {
+        Err(MyErrors::IndexOutOfBounds)
+    }
+}
+
+fn array_iterator(input: &Value) -> Result<Box<dyn Iterator<Item = Value>>, MyErrors> {
+    if let Some(array) = input.as_array() {
+        let iter = array.clone().into_iter();
+        return Ok(Box::new(iter));
+    }
+
+    Err(MyErrors::ListNotFound)
+}
+
+pub fn pipe(input: &Value, needle: &str) -> Result<FilterResult, MyErrors> {
+    let mut current_input = input.clone(); // Clone to use it mutably
+    if needle.contains(" | ") {
+        let mut sub_needles = needle.split(" | ");
+
+        while let Some(sub_needle) = sub_needles.next() {
+            let filter_result = filter_input(&current_input.clone(), sub_needle)?;
+            current_input = match filter_result {
+                FilterResult::SingleValue(value) => value,
+                FilterResult::Iterator(iter) => {
+                    if let Some(next_needle) = sub_needles.next() {
+                        let mut filtered_results = Vec::new();
+
+                        for item in iter {
+                            match filter_input(&item.clone(), next_needle) {
+                                Ok(FilterResult::SingleValue(filtered_value)) => {
+                                    filtered_results.push(filtered_value);
+                                }
+                                Ok(FilterResult::Iterator(nested_iter)) => {
+                                    filtered_results.extend(nested_iter);
+                                }
+                                Err(e) => {
+                                    return Err(e);
+                                }
+                            }
+                        }
+
+                        Value::Array(filtered_results)
+                    } else {
+                        return Ok(FilterResult::Iterator(iter));
+                    }
+                }
+            };
+        }
+
+        Ok(FilterResult::SingleValue(current_input))
+    } else {
+        let filter_result = filter_input(&current_input.clone(), needle)?;
+        Ok(filter_result)
+    }
+}
+
+pub enum FilterResult {
+    SingleValue(Value),
+    Iterator(Box<dyn Iterator<Item = Value>>),
+}
+
+impl PartialEq for FilterResult {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (FilterResult::SingleValue(a), FilterResult::SingleValue(b)) => a == b,
+            (FilterResult::Iterator(_), FilterResult::Iterator(_)) => true, // Just compare the variants, not the content.
```

Comment:
why did you choose to implement eq for iterators this way rather than checking each element?